apps/emqx/src/emqx_packet.erl

```diff
@@ -617,32 +617,84 @@ format_password(undefined) -> "";
 format_password(<<>>) -> "";
 format_password(_Password) -> "******".
 
+format_payload_label(Payload, Type) ->
+    ["Payload=", format_payload(Payload, Type)].
+
 format_payload(_, hidden) ->
-    "Payload=******";
-format_payload(Payload, text) when ?MAX_PAYLOAD_FORMAT_LIMIT(Payload) ->
-    ["Payload=", unicode:characters_to_list(Payload)];
-format_payload(Payload, hex) when ?MAX_PAYLOAD_FORMAT_LIMIT(Payload) ->
-    ["Payload(hex)=", binary:encode_hex(Payload)];
-format_payload(<<Part:?TRUNCATED_PAYLOAD_SIZE/binary, _/binary>> = Payload, Type) ->
-    [
-        "Payload=",
-        format_truncated_payload(Part, byte_size(Payload), Type)
-    ].
+    "******";
+format_payload(<<>>, _) ->
+    "";
+format_payload(Payload, Type) when ?MAX_PAYLOAD_FORMAT_LIMIT(Payload) ->
+    %% under the 1KB limit
+    format_payload_limit(Type, Payload, size(Payload));
+format_payload(Payload, Type) ->
+    %% too long, truncate to 100B
+    format_payload_limit(Type, Payload, ?TRUNCATED_PAYLOAD_SIZE).
+
+format_payload_limit(Type, Payload, Limit) when size(Payload) > Limit ->
+    {Part, TruncatedBytes} = truncate_payload(Type, Limit, Payload),
+    case TruncatedBytes > 0 of
+        true ->
+            [do_format_payload(Type, Part), "...(", integer_to_list(TruncatedBytes), " bytes)"];
+        false ->
+            do_format_payload(Type, Payload)
+    end;
+format_payload_limit(Type, Payload, _Limit) ->
+    do_format_payload(Type, Payload).
 
-format_truncated_payload(Bin, Size, Type) ->
-    Bin2 =
-        case Type of
-            text -> Bin;
-            hex -> binary:encode_hex(Bin)
-        end,
-    unicode:characters_to_list(
-        [
-            Bin2,
-            "... The ",
-            integer_to_list(Size - ?TRUNCATED_PAYLOAD_SIZE),
-            " bytes of this log are truncated"
-        ]
-    ).
+do_format_payload(text, Bytes) ->
+    try
+        [_ | _] = unicode:characters_to_list(Bytes)
+    catch
+        _:_ ->
+            do_format_payload(hex, Bytes)
+    end;
+do_format_payload(hex, Bytes) ->
+    ["hex:", binary:encode_hex(Bytes)].
+
+truncate_payload(hex, Limit, Payload) ->
+    <<Part:Limit/binary, Rest/binary>> = Payload,
+    {Part, size(Rest)};
+truncate_payload(text, Limit, Payload) ->
+    truncate_utf8(Limit, Payload).
+
+truncate_utf8(Limit, Payload) ->
+    CompleteLen = find_complete_utf8_len(Limit, Payload),
+    <<Part:CompleteLen/binary, Rest/binary>> = Payload,
+    {Part, size(Rest)}.
```

Comment:
Nit. Reimplementing UTF-8 stuff feels unnecessary to be honest, maybe something like this? Unless I'm missing some important aspect.

```suggestion
truncate_payload(text, Limit, Payload) ->
    case truncate_utf8(0, Limit, Payload) of
        Rest ->
            NRest = byte_size(Rest),
            NBytes = byte_size(Payload) - NRest,
            <<Part:NBytes, _/binary>> = Payload,
            {text, Part, NRest};
        false ->
            truncate_payload(hex, Limit, Payload)
    end.

truncate_utf8(Limit, <<_C/utf8, Rest/bytes>>) ->
    truncate_utf8(Limit - 1, Rest);
truncate_utf8(Limit, Rest) when Limit =< 0 ->
    Rest;
truncate_utf8(_Limit, _NonUtf8) ->
    false. %% Or return `_NonUtf8`?
```

Reply:
the idea is to avoid logging `hex` format at best-effort when opted to log `text`